In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:addyourpassword@localhost:XYZ/global_superstore

Connecting to 'mysql+pymysql://root:***@localhost:3306/global_superstore'

Let have a preview of our data

In [4]:
%%sql
SELECT
*
FROM
Superstore
LIMIT 4;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

4 rows affected.

+--------+----------------+------------+------------+----------------+-------------+----------------+----------+---------------+------------------+----------+-------------+--------+--------+-----------------+-----------------+--------------+------------------------------------+---------+----------+----------+----------+---------------+----------------+--------------+
| Row_ID |    Order_ID    | Order_Date | Ship_Date  |   Ship_Mode    | Customer_ID | Customer_Name  | Segment  |      City     |      State       | Country  | Postal_Code | Market | Region |    Product_ID   |     Category    | Sub_Category |            Product_Name            |  Sales  | Quantity | Discount |  Profit  | Shipping_Cost | Order_Priority | product_type |
+--------+----------------+------------+------------+----------------+-------------+----------------+----------+---------------+------------------+----------+-------------+--------+--------+-----------------+-----------------+--------------+------------------------------------+---------+----------+----------+----------+---------------+----------------+--------------+
|   1    | MX-2014-143658 | 2014-10-02 | 2014-10-06 | Standard Class |   SC-20575  |  Sonia Cooley  | Consumer |  Mexico City  | Distrito Federal |  Mexico  |             | LATAM  | North  | OFF-LA-10002782 | Office Supplies |    Labels    | Hon File Folder Labels, Adjustable |  13.080 |    3     |   0.00   | 4.56000  |    1.03000    |        Medium
         |  Adjustable  |
|   2    | MX-2012-155047 | 2012-10-15 | 2012-10-20 | Standard Class |   KW-16570  | Kelly Williams | Consumer | Dos Quebradas |    Risaralda     | Colombia |             | LATAM  | South  | FUR-FU-10004015 |    Furniture    | Furnishings  |        Tenex Clock, Durable        | 252.160 |    8     |   0.00   | 90.72000 |    13.45000   |        Medium
         |   Durable    |
|   3    | MX-2012-155047 | 2012-10-15 | 2012-10-20 | Standard Class |   KW-16570  | Kelly Williams | Consumer | Dos Quebradas |    Risaralda     | Colombia |             | LATAM  | South  | FUR-BO-10002352 |    Furniture    |  Bookcases   |    Ikea 3-Shelf Cabinet, Mobile    | 193.280 |    2     |   0.00   | 54.08000 |    9.63000    |        Medium
         |    Mobile    |
|   4    | MX-2012-155047 | 2012-10-15 | 2012-10-20 | Standard Class |   KW-16570  | Kelly Williams | Consumer | Dos Quebradas |    Risaralda     | Colombia |             | LATAM  | South  | OFF-BI-10004428 | Office Supplies |   Binders    |       Cardinal Binder, Clear       |  35.440 |    4     |   0.00   | 4.96000  |    1.37000    |        Medium
         |    Clear     |
+--------+----------------+------------+------------+----------------+-------------+----------------+----------+---------------+------------------+----------+-------------+--------+--------+-----------------+-----------------+--------------+------------------------------------+---------+----------+----------+----------+---------------+----------------+--------------+

In [6]:
%%sql
SELECT
*
FROM
Superstore
WHERE Product_name LIKE ("%Tenex Clock%")
    OR Product_name LIKE("%Sanford Canvas%")
LIMIT 4;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

4 rows affected.

+--------+----------------+------------+------------+----------------+-------------+----------------+-----------+---------------+------------+----------+-------------+--------+-----------+-----------------+-----------------+--------------+-----------------------------+---------+----------+----------+-----------+---------------+----------------+
| Row_ID |    Order_ID    | Order_Date | Ship_Date  |   Ship_Mode    | Customer_ID | Customer_Name  |  Segment  |      City     |   State    | Country  | Postal_Code | Market |   Region  |    Product_ID   |     Category    | Sub_Category |         Product_Name        |  Sales  | Quantity | Discount |   Profit  | Shipping_Cost | Order_Priority |
+--------+----------------+------------+------------+----------------+-------------+----------------+-----------+---------------+------------+----------+-------------+--------+-----------+-----------------+-----------------+--------------+-----------------------------+---------+----------+----------+-----------+---------------+----------------+
|   2    | MX-2012-155047 | 2012-10-15 | 2012-10-20 | Standard Class |   KW-16570  | Kelly Williams |  Consumer | Dos Quebradas | Risaralda  | Colombia |             | LATAM  |   South   | FUR-FU-10004015 |    Furniture    | Furnishings  |     Tenex Clock, Durable    | 252.160 |    8     |   0.00   |  90.72000 |    13.45000   |        Medium
         |
|   5    | MX-2012-155047 | 2012-10-15 | 2012-10-20 | Standard Class |   KW-16570  | Kelly Williams |  Consumer | Dos Quebradas | Risaralda  | Colombia |             | LATAM  |   South   | OFF-AR-10004594 | Office Supplies |     Art      | Sanford Canvas, Water Color |  71.600 |    2     |   0.00   |  11.44000 |    3.79000    |        Medium
         |
|  160   | MX-2014-133053 | 2014-12-18 | 2014-12-20 |  Second Class  |   GP-14740  |  Guy Phonely   | Corporate |   Cienfuegos  | Cienfuegos |   Cuba   |             | LATAM  | Caribbean | OFF-AR-10001694 | Office Supplies |     Art      |     Sanford Canvas, Blue    | 168.400 |    5     |   0.00   |  3.30000  |    24.44000   |        High
         |
|  384   | US-2012-148425 | 2012-06-21 | 2012-06-28 | Standard Class |   AB-10105  | Adrian Barton  |  Consumer |      Puno     |    Puno    |   Peru   |             | LATAM  |   South   | OFF-AR-10002335 | Office Supplies |     Art      | Sanford Canvas, Fluorescent |  62.136 |    3     |   0.40   | -39.38400 |    9.05000    |        Low
         |
+--------+----------------+------------+------------+----------------+-------------+----------------+-----------+---------------+------------+----------+-------------+--------+-----------+-----------------+-----------------+--------------+-----------------------------+---------+----------+----------+-----------+---------------+----------------+

From what we can see under the product name, there is a product type at the end separated by comma, we have to split these two by creating a new column

Let create a column of where we will keep the data 

In [36]:
%%sql
ALTER TABLE Superstore
ADD COLUMN product_type VARCHAR(40);--we will keep the product type in this new column--


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

RuntimeError: (pymysql.err.OperationalError) (1060, "Duplicate column name 'product_type'")
[SQL: ALTER TABLE Superstore
ADD COLUMN product_type VARCHAR(40);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [5]:
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

In [42]:
%%sql
UPDATE superstore
SET Product_type= 
    CASE 
    WHEN Product_name LIKE "%,%" THEN TRIM(SUBSTRING_INDEX(Product_name, ',', -1))
         ELSE NULL
    END;-- this one will look for the string that comes after , and we will set it in the new product type column--

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

51290 rows affected.

++
||
++
++

In [33]:
%%sql
ALTER TABLE Superstore
ADD COLUMN product VARCHAR(160);

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

In [34]:
%%sql
UPDATE superstore
SET Product= 
    CASE 
    WHEN Product_name LIKE "%,%" THEN TRIM(SUBSTRING_INDEX(Product_name, ',', 1))
         ELSE NULL
    END;;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

51290 rows affected.

RuntimeError: (pymysql.err.OperationalError) (1065, 'Query was empty')
[SQL: ;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


Let check for duplicates

In [6]:
%%sql
SELECT
Order_ID,
Order_Date,
Ship_Date,
Ship_Mode,
Customer_ID,
Customer_Name,
Segment,
City,
State,
Country,
Postal_Code,
Market,
Region,
COUNT(*) AS duplicate_count
FROM 
superstore
GROUP BY Order_ID,Order_Date,Ship_Date,Ship_Mode,Customer_ID,Customer_Name,Segment,City,State,Country,Postal_Code,Market,Region,Product_Name,Sales,Quantity,Discount,Profit,Shipping_Cost,Order_Priority
HAVING COUNT(*)>1
ORDER BY duplicate_count DESC
LIMIT 3;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

+----------+------------+-----------+-----------+-------------+---------------+---------+------+-------+---------+-------------+--------+--------+-----------------+
| Order_ID | Order_Date | Ship_Date | Ship_Mode | Customer_ID | Customer_Name | Segment | City | State | Country | Postal_Code | Market | Region | duplicate_count |
+----------+------------+-----------+-----------+-------------+---------------+---------+------+-------+---------+-------------+--------+--------+-----------------+
+----------+------------+-----------+-----------+-------------+---------------+---------+------+-------+---------+-------------+--------+--------+-----------------+

As we can see there are no duplicate entries in our table so let proceed

LET CREATE TABLES THAT WILL HELP US IN DATA MODELING

Let create new tables, that we will use to store different sales details.

1. Products table

In [36]:
%%sql
CREATE TABLE Product(
    Product_ID VARCHAR(20) PRIMARY KEY,
    Product_name VARCHAR(160),
    Product VARCHAR(80),
    Product_type VARCHAR(80),
    Category VARCHAR(50),
    Sub_category VARCHAR(50)
    )
;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

Let populate this table

In [45]:
%%sql
INSERT INTO product (
    product_id,
    product_name,
    category,
    sub_category,
    product,
    product_type
)
SELECT
    s.Product_ID,
    MAX(s.Product_Name)  AS product_name, --collapses attribute variations
    MAX(s.Category)      AS category,
    MAX(s.Sub_Category)  AS sub_category,
    MAX(s.product)       AS product,
    MAX(s.product_type)  AS product_type
FROM Superstore s
LEFT JOIN product p --prevents re-inserts
    ON s.Product_ID = p.product_id
WHERE s.Product_ID IS NOT NULL
  AND p.product_id IS NULL
GROUP BY s.Product_ID; --1 row per product


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

2. Customer table

In [47]:
%%sql
CREATE TABLE Customer (
    customer_id VARCHAR(30) PRIMARY KEY,
    customer_name VARCHAR(100),
    segment VARCHAR(50)
);

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

Let populate this

In [46]:
%%sql
INSERT INTO Customer(Customer_id,Customer_name,segment)
SELECT DISTINCT
    Customer_id,
    customer_name,
    segment
FROM Superstore;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1590 rows affected.

++
||
++
++

3. Location Table

In [48]:
%%sql 
CREATE TABLE Location(
location_id VARCHAR(15) AUTO INCREMENT Primary Key,
city VARCHAR(20),
state VARCHAR(20),
country VARCHAR(20),
postal_code INT,
market VARCHAR(20),
region VARCHAR(20)
)
;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

RuntimeError: (pymysql.err.OperationalError) (1050, "Table 'location' already exists")
[SQL: CREATE TABLE Location(
location_id VARCHAR(15) Primary Key,
city VARCHAR(20),
state VARCHAR(20),
country VARCHAR(20),
postal_code INT,
market VARCHAR(20),
region VARCHAR(20)
)
;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [52]:
%%sql
INSERT INTO Location(city,state,country,postal_code,market,region)
SELECT DISTINCT
city,
state,
country,
postal_code,
market,
region
FROM
superstore;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

3847 rows affected.

++
||
++
++

4. Date table

In [9]:
%%sql
CREATE TABLE Date(
date_id DATE Primary Key,
year INT,
quarter VARCHAR(10),
month INT,
month_name VARCHAR(10),
day VARCHAR(10)
)
;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

let populate the date table with data from ship date and order date columns from superstore table

In [18]:
%%sql
INSERT INTO date (date_id, year, quarter, month, month_name, day, day_of_week)
SELECT DISTINCT
    Order_Date AS date_id,
    YEAR(Order_Date),
    QUARTER(Order_Date),
    MONTH(Order_Date),
    MONTHNAME(Order_Date),
    DAY(Order_Date),
    DAYNAME(Order_Date)
FROM superstore
WHERE Order_Date IS NOT NULL;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1430 rows affected.

++
||
++
++

In [19]:
%%sql
INSERT IGNORE INTO date (date_id, year, quarter, month, month_name, day, day_of_week) --this IGNORE will help us to skip any existing data hence preventing the duplicates in our table
SELECT DISTINCT
    Ship_Date AS date_id,
    YEAR(Ship_Date),
    QUARTER(Ship_Date),
    MONTH(Ship_Date),
    MONTHNAME(Ship_Date),
    DAY(Ship_Date),
    DAYNAME(Ship_Date)
FROM superstore
WHERE Ship_Date IS NOT NULL;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

38 rows affected.

++
||
++
++

5. Shipping Table

In [10]:
%%sql
CREATE TABLE Shipping(
ship_mode_id VARCHAR(10) AUTO INCREMENT Primary Key,
ship_mode VARCHAR(20)
)
;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

In [ ]:
Let populate this table

In [53]:
%%sql
INSERT INTO shipping(ship_mode)
SELECT DISTINCT 
ship_mode
FROM
superstore;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

4 rows affected.

++
||
++
++

6. Fact table


this table will syore measurable business events

In [16]:
%%sql
CREATE TABLE fact_sales (
    sales_id INT AUTO_INCREMENT PRIMARY KEY,
    order_id VARCHAR(20) NOT NULL,
    customer_id VARCHAR(20) NOT NULL,
    product_id VARCHAR(20) NOT NULL,
    location_id INT NOT NULL,
    order_date_id DATE NOT NULL,
    ship_date_id DATE NOT NULL,
    ship_mode_id INT NOT NULL,
    sales DECIMAL(10,3) NOT NULL,
    quantity INT NOT NULL,
    discount DECIMAL(4,2) NOT NULL,
    profit DECIMAL(10,5) NOT NULL,
    shipping_cost DECIMAL(10,5) NOT NULL,
    order_priority VARCHAR(20),

    --constraints
    FOREIGN KEY (customer_id) REFERENCES customer(customer_id),
    FOREIGN KEY (product_id) REFERENCES product(product_id),
    FOREIGN KEY (location_id) REFERENCES location(location_id),
    FOREIGN KEY (order_date_id) REFERENCES date(date_id),
    FOREIGN KEY (ship_date_id) REFERENCES date(date_id),
    FOREIGN KEY (ship_mode_id) REFERENCES shipping(ship_mode_id)
);


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

Let populate the fact table

In [56]:
%%sql
INSERT INTO Fact_sales(
order_id ,
customer_id ,
product_id ,
location_id ,
order_date_id ,
ship_date_id  ,
ship_mode_id ,
sales ,
quantity ,
discount ,
profit ,
shipping_cost ,
order_priority)
SELECT
    ss.Order_ID,
    ss.Customer_ID,
    ss.Product_ID,
    l.location_id,
    ss.Order_Date,
    ss.Ship_Date,
    s.ship_mode_id,
    ss.Sales,
    ss.Quantity,
    ss.Discount,
    ss.Profit,
    ss.Shipping_Cost,
    ss.Order_Priority
FROM superstore ss
JOIN customer c
    ON ss.Customer_ID = c.customer_id
JOIN product p
    ON ss.Product_ID = p.product_id
JOIN location l
    ON ss.City = l.city
   AND ss.Country = l.country
   AND (ss.State = l.state OR ss.State IS NULL)
JOIN shipping s
    ON ss.Ship_Mode = s.ship_mode
JOIN date d
    ON ss.Order_Date = d.date_id
JOIN date dd
    ON ss.Ship_Date = dd.date_id;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

64362 rows affected.

++
||
++
++

In [62]:
%%sql
SELECT COUNT(*) FROM superstore;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51290   |
+----------+

In [61]:
%%sql
SELECT COUNT(*) FROM fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  64362   |
+----------+

The fact_sales has more rows than our original table(superstore), this signals that there are duplicates created hence resulting in extra columns,let look on how we can spot the issue
the issue remains but the number is decreasing, let look for another source of the issue
||Now:
rows in superstore=51290
rows in fact_sales=64362

Let start by checking duplicates in location table 

In [63]:
%%sql
SELECT 
    city,
    state,
    country,
    COUNT(*) AS cnt
FROM location
GROUP BY city, state, country
HAVING cnt > 1;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

18 rows affected.

+---------------+---------------+---------------+-----+
|      city     |     state     |    country    | cnt |
+---------------+---------------+---------------+-----+
|     Vienna    |     Vienna    |    Austria    |  2  |
|      Linz     | Upper Austria |    Austria    |  2  |
|    Salzburg   |    Salzburg   |    Austria    |  2  |
|      Graz     |     Styria    |    Austria    |  2  |
|   Innsbruck   |     Tyrol     |    Austria    |  2  |
|   Klagenfurt  |   Carinthia   |    Austria    |  2  |
|   Ulan Bator  |  Ulaanbaatar  |    Mongolia   |  2  |
|  Los Angeles  |   California  | United States |  6  |
|    Seattle    |   Washington  | United States |  3  |
| San Francisco |   California  | United States |  3  |
|  Philadelphia |  Pennsylvania | United States |  4  |
|    Houston    |     Texas     | United States |  4  |
| New York City |    New York   | United States |  4  |
|    Chicago    |    Illinois   | United States |  3  |
|     Dallas    |     Texas     | United States |  3  |
|    Detroit    |    Michigan   | United States |  3  |
|   San Diego   |   California  | United States |  3  |
|     Miami     |    Florida    | United States |  3  |
+---------------+---------------+---------------+-----+

those are duplicates 

In [70]:
%%sql
SELECT ss.Row_ID, COUNT(*) AS join_count
FROM superstore ss
JOIN location l
  ON ss.City = l.city
 AND ss.Country = l.country
GROUP BY ss.Row_ID
HAVING join_count > 1
LIMIT 10;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

10 rows affected.

+--------+------------+
| Row_ID | join_count |
+--------+------------+
| 10370  |     2      |
| 10371  |     2      |
| 10372  |     2      |
| 10582  |     2      |
| 10623  |     2      |
| 10692  |     2      |
| 10693  |     2      |
| 10702  |     2      |
| 10788  |     2      |
| 10919  |     2      |
+--------+------------+

ALL THE ABOVE QUERRIES RETURNING THOSE ID MEANS each one is being duplicated

LET FIX THESE ISSUES

In [ ]:
%%sql
ALTER TABLE location
ADD CONSTRAINT un_location
UNIQUE (city, state, country, postal_code, market, region);--Enforce Uniqueness


In [78]:
%%sql
TRUNCATE TABLE Fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

In [83]:
%%sql
INSERT INTO fact_sales (
    order_id,
    customer_id,
    product_id,
    location_id,
    order_date_id,
    ship_date_id,
    ship_mode_id,
    sales,
    quantity,
    discount,
    profit,
    shipping_cost,
    order_priority
)
SELECT
    ss.Order_ID,
    ss.Customer_ID,
    ss.Product_ID,
    l.location_id,
    ss.Order_Date,
    ss.Ship_Date,
    sh.ship_mode_id,
    ss.Sales,
    ss.Quantity,
    ss.Discount,
    ss.Profit,
    ss.Shipping_Cost,
    ss.Order_Priority
FROM superstore ss
JOIN customer c
    ON ss.Customer_ID = c.customer_id
JOIN product p
    ON ss.Product_ID = p.product_id
JOIN location l
    ON ss.City = l.city
   AND ss.Country = l.country
   AND ss.State = l.state
   AND (ss.Postal_Code = l.postal_code OR ss.Postal_Code IS NULL)--a fix was made on the location join
JOIN shipping sh
    ON ss.Ship_Mode = sh.ship_mode
JOIN date d
    ON ss.Order_Date = d.date_id
JOIN date dd
    ON ss.Ship_Date = dd.date_id;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

51638 rows affected.

++
||
++
++

In [81]:
%%sql
SELECT COUNT(*) FROM superstore;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51290   |
+----------+

In [82]:
%%sql
SELECT COUNT(*) FROM fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51638   |
+----------+

the issue remains but the number is decreasing, let look for another source of the issue
||Now:
rows in superstore=51290
rows in superstore=51638

In [84]:
%%sql
SELECT ss.Row_ID, COUNT(*) AS matches
FROM superstore ss
JOIN location l
  ON ss.City = l.city
 AND ss.Country = l.country
 AND ss.State = l.state
GROUP BY ss.Row_ID
HAVING COUNT(*) > 1;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

4675 rows affected.

+--------+---------+
| Row_ID | matches |
+--------+---------+
| 10370  |    2    |
| 10371  |    2    |
| 10372  |    2    |
| 10582  |    2    |
| 10692  |    2    |
| 10693  |    2    |
| 10702  |    2    |
| 10788  |    2    |
| 10919  |    2    |
| 10963  |    2    |
| 11211  |    2    |
| 11212  |    2    |
| 11315  |    2    |
| 11343  |    2    |
| 11344  |    2    |
| 11597  |    2    |
| 11656  |    2    |
| 11717  |    2    |
| 11750  |    2    |
| 11771  |    2    |
| 11787  |    2    |
| 11800  |    2    |
| 11807  |    2    |
| 12035  |    2    |
| 12036  |    2    |
| 12037  |    2    |
| 12099  |    2    |
| 12239  |    2    |
| 12240  |    2    |
| 12645  |    2    |
| 12727  |    2    |
| 12764  |    2    |
| 12788  |    2    |
| 12829  |    2    |
| 12830  |    2    |
| 12831  |    2    |
| 12840  |    2    |
| 12906  |    2    |
| 12947  |    2    |
| 13138  |    2    |
| 13139  |    2    |
| 13140  |    2    |
| 13141  |    2    |
| 13142  |    2    |
| 13195  |    2    |
| 13237  |    2    |
| 13238  |    2    |
| 13239  |    2    |
| 13240  |    2    |
| 13241  |    2    |
| 13242  |    2    |
| 13300  |    2    |
| 13327  |    2    |
| 13353  |    2    |
| 13354  |    2    |
| 13362  |    2    |
| 13371  |    2    |
| 13403  |    2    |
| 13404  |    2    |
| 13405  |    2    |
| 13483  |    2    |
| 13484  |    2    |
| 13516  |    2    |
| 13520  |    2    |
| 13521  |    2    |
| 13522  |    2    |
| 13707  |    2    |
| 13708  |    2    |
| 13709  |    2    |
| 13854  |    2    |
| 14024  |    2    |
| 14025  |    2    |
| 14026  |    2    |
| 14157  |    2    |
| 14158  |    2    |
| 14273  |    2    |
| 14274  |    2    |
| 14275  |    2    |
| 14276  |    2    |
| 14277  |    2    |
| 14278  |    2    |
| 14279  |    2    |
| 14297  |    2    |
| 14298  |    2    |
| 14299  |    2    |
| 14325  |    2    |
| 14331  |    2    |
| 14332  |    2    |
| 14407  |    2    |
| 14503  |    2    |
| 14504  |    2    |
| 14505  |    2    |
| 14571  |    2    |
| 14572  |    2    |
| 14578  |    2    |
| 14579  |    2    |
| 14716  |    2    |
| 14717  |    2    |
| 14718  |    2    |
| 14719  |    2    |
| 14778  |    2    |
| 14781  |    2    |
| 14827  |    2    |
| 14828  |    2    |
| 14881  |    2    |
| 15030  |    2    |
| 15031  |    2    |
| 15119  |    2    |
| 15148  |    2    |
| 15280  |    2    |
| 15281  |    2    |
| 15282  |    2    |
| 15339  |    2    |
| 15340  |    2    |
| 15341  |    2    |
| 15391  |    2    |
| 15465  |    2    |
| 15607  |    2    |
| 15656  |    2    |
| 15657  |    2    |
| 15694  |    2    |
| 15695  |    2    |
| 15742  |    2    |
| 15827  |    2    |
| 15828  |    2    |
| 15829  |    2    |
| 15830  |    2    |
| 15831  |    2    |
| 15934  |    2    |
| 15935  |    2    |
| 15936  |    2    |
| 15937  |    2    |
| 15938  |    2    |
| 16037  |    2    |
| 16038  |    2    |
| 16039  |    2    |
| 16040  |    2    |
| 16041  |    2    |
| 16042  |    2    |
| 16043  |    2    |
| 16119  |    2    |
| 16120  |    2    |
| 16145  |    2    |
| 16370  |    2    |
| 16462  |    2    |
| 16468  |    2    |
| 16487  |    2    |
| 16545  |    2    |
| 16579  |    2    |
| 16632  |    2    |
| 16633  |    2    |
| 16634  |    2    |
| 16663  |    2    |
| 16705  |    2    |
| 16706  |    2    |
| 16707  |    2    |
| 16708  |    2    |
| 16709  |    2    |
| 16758  |    2    |
| 16759  |    2    |
| 16760  |    2    |
| 16761  |    2    |
| 16825  |    2    |
| 16826  |    2    |
| 16827  |    2    |
| 16904  |    2    |
| 16923  |    2    |
| 16950  |    2    |
| 16951  |    2    |
| 17041  |    2    |
| 17042  |    2    |
| 17076  |    2    |
| 17139  |    2    |
| 17230  |    2    |
| 17272  |    2    |
| 17273  |    2    |
| 17380  |    2    |
| 17396  |    2    |
| 17458  |    2    |
| 17459  |    2    |
| 17460  |    2    |
| 17461  |    2    |
| 17587  |    2    |
| 17588  |    2    |
| 17596  |    2    |
| 17637  |    2    |
| 17638  |    2    |
| 17746  |

Let create a clean location table, that has no dupplicates

In [85]:
%%sql
CREATE TABLE location_clean AS
SELECT
    MIN(location_id) AS location_id,
    city,
    state,
    country
FROM location
GROUP BY city, state, country;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

3812 rows affected.

++
||
++
++

In [88]:
%%sql
TRUNCATE TABLE fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

In [89]:
%%sql
INSERT INTO fact_sales (
    order_id,
    customer_id,
    product_id,
    location_id,
    order_date_id,
    ship_date_id,
    ship_mode_id,
    sales,
    quantity,
    discount,
    profit,
    shipping_cost,
    order_priority
)
SELECT
    ss.Order_ID,
    ss.Customer_ID,
    ss.Product_ID,
    lc.location_id,
    ss.Order_Date,
    ss.Ship_Date,
    sh.ship_mode_id,
    ss.Sales,
    ss.Quantity,
    ss.Discount,
    ss.Profit,
    ss.Shipping_Cost,
    ss.Order_Priority
FROM superstore ss
JOIN customer c
    ON ss.Customer_ID = c.customer_id
JOIN product p
    ON ss.Product_ID = p.product_id
JOIN location_clean lc
    ON ss.City = lc.city
   AND ss.Country = lc.country
   AND ss.State = lc.state --a fix was made on the location join
JOIN shipping sh
    ON ss.Ship_Mode = sh.ship_mode
JOIN date d
    ON ss.Order_Date = d.date_id
JOIN date dd
    ON ss.Ship_Date = dd.date_id;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

51290 rows affected.

++
||
++
++

In [90]:
%%sql
SELECT COUNT(*) FROM superstore;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51290   |
+----------+

In [91]:
%%sql
SELECT COUNT(*) FROM fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51290   |
+----------+

now that we got count_superstore=count_fact_sales 
the issue has been solved there are no lost data and no duplicates in the fact_sales table

Let create a unique one table for location and delete other location tables that contained duplicates

In [93]:
%%sql
CREATE TABLE location_cleaned AS
SELECT
    MIN(location_id) AS location_id,  -- keeps one surrogate key per group
    city,
    state,
    country,
    MIN(postal_code) AS postal_code,  -- pick one postal code per location
    MIN(market) AS market,
    MIN(region) AS region
FROM location
GROUP BY city, state, country;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

3812 rows affected.

++
||
++
++

In [94]:
%%sql
SELECT city, state, country, COUNT(*)
FROM location_clean
GROUP BY city, state, country
HAVING COUNT(*) > 1;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

+------+-------+---------+----------+
| city | state | country | COUNT(*) |
+------+-------+---------+----------+
+------+-------+---------+----------+

In [97]:
%%sql
DROP TABLE location;

ALTER TABLE location_cleaned RENAME TO location;


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

In [105]:
%%sql
TRUNCATE TABLE fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++

In [106]:
%%sql
INSERT INTO fact_sales (
    order_id,
    customer_id,
    product_id,
    location_id,
    order_date_id,
    ship_date_id,
    ship_mode_id,
    sales,
    quantity,
    discount,
    profit,
    shipping_cost,
    order_priority
)
SELECT
    ss.Order_ID,
    ss.Customer_ID,
    ss.Product_ID,
    l.location_id,
    ss.Order_Date,
    ss.Ship_Date,
    sh.ship_mode_id,
    ss.Sales,
    ss.Quantity,
    ss.Discount,
    ss.Profit,
    ss.Shipping_Cost,
    ss.Order_Priority
FROM superstore ss
JOIN customer c
    ON ss.Customer_ID = c.customer_id
JOIN product p
    ON ss.Product_ID = p.product_id
JOIN location l
    ON ss.City = l.city
   AND ss.Country = l.country
   AND ss.State = l.state --a fix was made on the location join
JOIN shipping sh
    ON ss.Ship_Mode = sh.ship_mode
JOIN date d
    ON ss.Order_Date = d.date_id
JOIN date dd
    ON ss.Ship_Date = dd.date_id;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

51290 rows affected.

++
||
++
++

In [107]:
%%sql
SELECT
COUNT(*)
FROM Superstore;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51290   |
+----------+

In [108]:
%%sql
SELECT
COUNT(*)
FROM fact_sales;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

1 rows affected.

+----------+
| COUNT(*) |
+----------+
|  51290   |
+----------+

In [111]:
%%sql
SELECT ss.Row_ID, COUNT(*) AS matches
FROM superstore ss
JOIN location l
  ON ss.City = l.city
 AND ss.Country = l.country
 AND ss.State = l.state
GROUP BY ss.Row_ID
HAVING COUNT(*) > 1;
--this verifies if there are no duplicates remaining 


Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

+--------+---------+
| Row_ID | matches |
+--------+---------+
+--------+---------+

In [109]:
%%sql
DROP TABLE location_clean;

Running query in 'mysql+pymysql://root:***@localhost:3306/global_superstore'

++
||
++
++